src/optimizer/stats/tuple_sampler.cpp

```diff
@@ -0,0 +1,126 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// tuple_sampler.cpp
+//
+// Identification: src/optimizer/tuple_sampler.cpp
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "optimizer/stats/tuple_sampler.h"
+
+#include "storage/data_table.h"
+#include "storage/tile_group.h"
+#include "storage/tile_group_header.h"
+#include "storage/tile.h"
+#include "storage/tuple.h"
+
+namespace peloton {
+namespace optimizer {
+
+/**
+ * AcquireSampleTuples - Sample a certain number of tuples from a given table.
+ * This function performs random sampling by generating random tile_group_offset
+ * and random tuple_offset.
+ */
+size_t TupleSampler::AcquireSampleTuples(size_t target_sample_count) {
+  size_t tuple_count = table->GetTupleCount();
+  size_t tile_group_count = table->GetTileGroupCount();
+  LOG_TRACE("tuple_count = %lu, tile_group_count = %lu", tuple_count,
+            tile_group_count);
+
+  if (tuple_count < target_sample_count) {
+    target_sample_count = tuple_count;
+  }
+  size_t sampled_count = 0;
+
+  size_t rand_tilegroup_offset, rand_tuple_offset;
+  srand(time(NULL));
+  catalog::Schema *tuple_schema = table->GetSchema();
+
+  while (sampled_count < target_sample_count) {
+    // Generate a random tilegroup offset
+    rand_tilegroup_offset = rand() % tile_group_count;
+    storage::TileGroup *tile_group =
+        table->GetTileGroup(rand_tilegroup_offset).get();
+    oid_t tuple_per_group = tile_group->GetActiveTupleCount();
+    LOG_TRACE("tile_group: offset: %lu, addr: %p, tuple_per_group: %u",
+              rand_tilegroup_offset, tile_group, tuple_per_group);
+    if (tuple_per_group == 0) {
+      continue;
+    }
+
+    rand_tuple_offset = rand() % tuple_per_group;
+
+    std::unique_ptr<storage::Tuple> tuple(
+        new storage::Tuple(tuple_schema, true));
+
+    LOG_TRACE("tuple_group_offset = %lu, tuple_offset = %lu",
+              rand_tilegroup_offset, rand_tuple_offset);
+    if (!GetTupleInTileGroup(tile_group, rand_tuple_offset, tuple)) {
+      continue;
+    }
+    LOG_TRACE("Add sampled tuple: %s", tuple->GetInfo().c_str());
+    sampled_tuples.push_back(std::move(tuple));
+    sampled_count++;

```

Comment:
Why is there a need for doing this ? The while loop in line 44 could be written as `while (sampled_count.size() < target_sample_count)` . size() will likely be inlined by the compiler to directly access the private size field on the vector<T> class.